pinot-core/src/main/java/org/apache/pinot/core/common/ColumnValueReader.java

```diff
@@ -0,0 +1,248 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.common;
+
+import org.apache.pinot.core.io.reader.ForwardIndexReader;
+import org.apache.pinot.core.io.reader.ReaderContext;
+import org.apache.pinot.spi.data.FieldSpec.DataType;
+
+
+/**
+ * ColumnValueReader is a wrapper class on top of the forward index reader to read values from it.
+ * <p>Batch value read APIs support the following data type conversions for numeric types:
+ * <ul>
+ *   <li>INT -> LONG</li>
+ *   <li>INT, LONG -> FLOAT</li>
+ *   <li>INT, LONG, FLOAT -> DOUBLE</li>
+ * </ul>
+ */
+@SuppressWarnings({"rawtypes", "unchecked"})
+public final class ColumnValueReader {
+  private final ForwardIndexReader _forwardIndexReader;
+  private final DataType _valueType;
+  private final ReaderContext _readerContext;
+
+  public ColumnValueReader(ForwardIndexReader forwardIndexReader) {
+    _forwardIndexReader = forwardIndexReader;
+    _valueType = forwardIndexReader.getValueType();
+    // TODO: Figure out a way to close the reader context. Currently it can cause direct memory leak.
```

Comment:
This should not be a TODO, but fixed in this PR itself? Or are you saying it happens in some corner case? If so, please describe when that might happen.

Reply:
This is the current behavior. The fix is not very straight forward, so I don't want to mix this fix with the refactor PR. Will try to fix it in the next PR.